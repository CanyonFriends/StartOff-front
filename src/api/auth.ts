import axios from 'axios';
import { SignupInfoType } from '../validator/signupValidator';
import { LoginInfoType } from '../validator/loginValidator';
import { ErrorType } from './error';

export interface SigninResponseType {
  access_token: string;
  uuid: string;
  email: string;
  nickname: string;
  user_id: string;
}

export interface SignupResponseType {
  user_id: string;
}

export interface LogoutRequestType {
  email: string;
  accessToken: string;
  uuid: string;
}

export interface LogoutResponseType {
  msg: string;
  success: boolean;
}

export const signupAPI = async ({ id, pw, nickname }: SignupInfoType): Promise<ErrorType | SignupResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/signup',
      data: {
        nickname,
        email: id,
        password: pw,
      },
    });
    return response.data as SignupResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};

export const signinAPI = async ({ id, pw }: LoginInfoType): Promise<ErrorType | SigninResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/login',
      data: {
        email: id,
        password: pw,
      },
    });
    return response.data as SigninResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};

export const logoutAPI = async ({
  email,
  accessToken,
  uuid,
}: LogoutRequestType): Promise<ErrorType | LogoutResponseType> => {
  try {
    const response = await axios({
      method: 'POST',
      url: '/v1/logout',
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
      data: {
        email,
        accessToken,
        uuid,
      },
    });
    return response.data as LogoutResponseType;
  } catch (error) {
    return { errorMsg: error.response.data.errorMsg } as ErrorType;
  }
};
