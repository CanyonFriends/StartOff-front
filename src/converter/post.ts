import { BoardServerType, SummarizedPostServerType, CreatePostServerType, PostServerType } from '../@types/server';
import { BoardClientType, SummarizedPostClientType, CreatePostClientType, PostClientType } from '../@types/client';
import { skillServerType2ClientType } from './skill';

export const summarizedPostServerType2ClientType = (
  summarizedPost: SummarizedPostServerType,
): SummarizedPostClientType => {
  return {
    postId: String(summarizedPost.post_id),
    title: summarizedPost.title,
    currentPeople: summarizedPost.current_people,
    maxPeople: summarizedPost.max_people,
    createAt: new Date(summarizedPost.created_at),
    nickname: summarizedPost.nickname,
    postSkills: summarizedPost.post_skills.map((skill) => skillServerType2ClientType(skill)),
  };
};

export const boardServerType2ClientType = (board: BoardServerType): BoardClientType => {
  return {
    content: board.content.map((post) => summarizedPostServerType2ClientType(post)),
    totalElements: board.totalElements,
    totalPages: board.totalPages,
  };
};

export const createPostClientType2ServerType = (post: CreatePostClientType): CreatePostServerType => {
  return {
    category: post.category.toUpperCase(),
    content: post.content,
    current_people: post.currentPeople,
    max_people: post.maxPeople,
    post_skills: post.postSkills.map((skill) => skill.skillName),
    title: post.title,
    user_id: Number(post.userId),
  };
};

export const postServerType2ClientType = (post: PostServerType): PostClientType => {
  return {
    postId: String(post.post_id),
    category: post.category,
    title: post.title,
    content: post.content,
    createdAt: new Date(post.created_at),
    currentPeople: post.current_people,
    maxPeople: post.max_people,
    nickname: post.nickname,
    postSkills: post.post_skills.map((skill) => skillServerType2ClientType(skill)),
    userId: String(post.user_id),
  };
};
