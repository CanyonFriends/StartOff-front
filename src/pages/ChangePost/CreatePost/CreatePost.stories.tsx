import React from 'react';
import { action } from '@storybook/addon-actions';
import { makeSkillMock } from '../../../__mocks__/client-mock-data';
import CreatePostTemplate from '../template';

export default {
  title: 'Template/CreatePost',
  component: CreatePostTemplate,
};

const totalSkillList = [
  makeSkillMock({ skillName: 'typescript' }),
  makeSkillMock({ skillName: 'javascript' }),
  makeSkillMock({ skillName: 'python' }),
];

const changeAction = action('onChange');
const changeAsyncAction = async () => {
  changeAction();
  return '';
};
export const createPostTemplate = (): React.ReactElement => (
  <CreatePostTemplate totalSkillList={totalSkillList} handleSubmit={changeAsyncAction} />
);
