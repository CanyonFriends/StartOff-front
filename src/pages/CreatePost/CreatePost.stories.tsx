import React from 'react';
import { makeSkillMock } from '../../__mocks__/client-mock-data';
import CreatePostTemplate from './template';

export default {
  title: 'Template/CreatePost',
  component: CreatePostTemplate,
};

const totalSkillList = [
  makeSkillMock({ skillName: 'typescript' }),
  makeSkillMock({ skillName: 'javascript' }),
  makeSkillMock({ skillName: 'python' }),
];
export const createPostTemplate = (): React.ReactElement => (
  <CreatePostTemplate totalSkillList={totalSkillList} board="board" userId="1" />
);
