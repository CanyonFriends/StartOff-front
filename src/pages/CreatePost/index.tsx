import React from 'react';
import CreatePostTemplate from './template';

function CreatePost() {
  return <CreatePostTemplate />;
}

export default CreatePost;
