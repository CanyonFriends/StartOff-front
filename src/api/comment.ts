import { AxiosError } from 'axios';
import { ErrorType } from './error';
import axios from '../utils/axios';
import { CommentServerType } from '../@types/server';
import { commentServerType2ClientType } from '../converter/comment';

export interface CreateCommentRequest {
  postId: string;
  userId: string;
  content: string;
}

export interface UpdateCommentRequest {
  postId: string;
  userId: string;
  content: string;
  commentId: string;
}

export interface DeleteCommentRequest {
  postId: string;
  commentId: string;
}

export const createCommentAPI = async ({ postId, userId, content }: CreateCommentRequest) => {
  try {
    const response = await axios({
      method: 'POST',
      url: `/v1/posts/${postId}/comments`,
      data: {
        parend_id: null,
        user_id: userId,
        content,
      },
    });

    return commentServerType2ClientType(response.data as CommentServerType);
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};

export const updateCommentAPI = async ({ postId, userId, content, commentId }: UpdateCommentRequest) => {
  try {
    const response = await axios({
      method: 'PUT',
      url: `/v1/posts/${postId}/comments/${commentId}`,
      data: {
        parend_id: null,
        user_id: userId,
        content,
      },
    });

    return commentServerType2ClientType(response.data as CommentServerType);
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};

export const deleteCommentAPI = async ({ postId, commentId }: DeleteCommentRequest) => {
  try {
    await axios({
      method: 'DELETE',
      url: `/v1/posts/${postId}/comments/${commentId}`,
    });

    return true;
  } catch (error) {
    return { error_msg: (error as AxiosError).response?.data?.error_msg } as ErrorType;
  }
};
