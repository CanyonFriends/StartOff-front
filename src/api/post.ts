import axios from '../utils/axios';
import { ErrorType } from './error';
import { BoardServerType } from '../@types/server';
import { boardServerType2ClientType, createPostClientType2ServerType } from '../converter/post';
import { CreatePostClientType } from '../@types/client';

export const getCategoriesAPI = async () => {
  try {
    const response = await axios({
      method: 'GET',
      url: '/v1/categories',
    });

    return response.data as string[];
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const getPostsAPI = async (page: number, size: number, category: string) => {
  try {
    const response = await axios({
      method: 'GET',
      url: `/v1/posts?page=${page}&size=${size}&category=${category}`,
    });

    return boardServerType2ClientType(response.data as BoardServerType);
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};

export const createPostAPI = async (post: CreatePostClientType) => {
  try {
    await axios({
      method: 'POST',
      url: '/v1/posts',
      data: {
        ...createPostClientType2ServerType(post),
      },
    });
    return true;
  } catch (error) {
    return { error_msg: error.response.data.error_msg } as ErrorType;
  }
};
