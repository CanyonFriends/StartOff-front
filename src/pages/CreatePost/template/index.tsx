import React, { useMemo } from 'react';
import { useHistory } from 'react-router-dom';
import * as Style from './styled';
import CommonHeader from '../../../components/Layout/CommonHeader';
import { Dropdown, InputField } from '../../../components/UI/molecule';
import PostMarkdownForm from '../../../components/UI/organism/PostMarkdownForm';
import useForm from '../../../hooks/useForm';
import postFormValidator, { PostFormValidatorType } from '../../../validator/postFormValidator';
import { InputFieldProps } from '../../../components/UI/molecule/InputField';
import { SkillList, AlertModal } from '../../../components/UI/organism';
import { SkillClientType } from '../../../@types/client';
import { Button, Title } from '../../../components/UI/atom';
import { createPostAPI } from '../../../api/post';
import { isFailed } from '../../../api/error';
import { buildBoardPath } from '../../../Routes';

interface CreatePostTemplateProps {
  totalSkillList: SkillClientType[];
  board: string;
  userId: string;
}

function CreatePostTemplate({ totalSkillList, board, userId }: CreatePostTemplateProps) {
  const history = useHistory();
  const createPostSubmit = async (values: PostFormValidatorType) => {
    const response = await createPostAPI({
      userId,
      category: board,
      content: values.content,
      currentPeople: values.currentPeople,
      maxPeople: values.maxPeople,
      title: values.title,
      postSkills: values.postSkills,
    });

    if (isFailed<boolean>(response)) {
      return response.error_msg;
    }

    history.push(buildBoardPath(board, 0));
    return '';
  };
  const { values, error, clearError, handleChange, handleSubmitWithErrorControl } = useForm<PostFormValidatorType>({
    initialState: {
      title: '',
      content: '',
      currentPeople: 0,
      maxPeople: 0,
      postSkills: [],
    },
    validator: postFormValidator,
    onSubmit: createPostSubmit,
  });
  const peopleNumber = useMemo(() => {
    return new Array(31).fill(0).map((_, index) => ({ id: String(index), text: String(index) }));
  }, []);

  const handleTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (event.target.value.length > 20) return;
    handleChange({ ...values, title: event.target.value });
  };

  const handleCurrentPeopleChange = (value: string) => {
    handleChange({ ...values, currentPeople: Number(value) });
  };

  const handleMaxPeopleChange = (value: string) => {
    handleChange({ ...values, maxPeople: Number(value) });
  };

  const handleContentChange = (content: string) => {
    handleChange({ ...values, content });
  };

  const clickTotalSkillItem = (skillName: string) => {
    const selectedItem = totalSkillList.find((skill) => skill.skillName === skillName);
    if (!selectedItem) return;
    handleChange({ ...values, postSkills: [...values.postSkills, selectedItem] });
  };

  const deleteMySkill = (skillId: string) => {
    const skillListExceptDeletedSkill = values.postSkills.filter((skill) => skill.skillId !== skillId);
    handleChange({ ...values, postSkills: skillListExceptDeletedSkill });
  };

  const titleInputField: InputFieldProps = {
    group: 'title',
    ariaLabel: 'post-title',
    labelProps: { content: '제목', size: 'extraLarge', bolder: true },
    inputProps: { value: values.title, onChange: handleTitleChange, placeholder: '제목을 입력하세요', size: 'large' },
    sortDirection: 'column',
  };
  return (
    <>
      {!!error && <AlertModal content={error} clickCloseButton={clearError} />}
      <CommonHeader />
      <Style.Container onSubmit={handleSubmitWithErrorControl}>
        <InputField {...titleInputField} />
        <Style.PeopleWrapper>
          <Title fontsize="h3">현재 인원 / 모집 인원</Title>
          <Style.DropdownWrapper>
            <Dropdown
              isClickValueText
              placeholder={`현재 인원 : ${values.currentPeople}`}
              items={peopleNumber}
              clickItem={handleCurrentPeopleChange}
            />
            <Dropdown
              isClickValueText
              placeholder={`모집 인원 : ${values.maxPeople}`}
              items={peopleNumber}
              clickItem={handleMaxPeopleChange}
            />
          </Style.DropdownWrapper>
        </Style.PeopleWrapper>
        <Style.SkillListWrapper>
          <SkillList
            editableAuthority
            title="기술 스택"
            mySkillList={values.postSkills}
            totalSkillList={totalSkillList}
            clickTotalSkillItem={clickTotalSkillItem}
            deleteMySkill={deleteMySkill}
          />
        </Style.SkillListWrapper>
        <Style.MarkdownWrapper>
          <PostMarkdownForm markdown={values.content} handleChangeMarkdownText={handleContentChange} />
        </Style.MarkdownWrapper>
        <Button size="large" width="100%">
          글 쓰기
        </Button>
      </Style.Container>
    </>
  );
}

export default CreatePostTemplate;
