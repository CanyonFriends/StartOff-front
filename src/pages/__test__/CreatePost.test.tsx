/**
 * @jest-environment jsdom
 */
import React from 'react';
import { waitFor } from '@testing-library/react';
import { render } from '../../test-utils';
import CreatePost from '../CreatePost';
import { getCategoriesAPI } from '../../api/post';

jest.mock('../../api/post');

const getCategoriesMockAPI = getCategoriesAPI as jest.MockedFunction<typeof getCategoriesAPI>;

beforeEach(() => {
  jest.resetAllMocks();
  getCategoriesMockAPI.mockReturnValue(new Promise((res) => res([])));
});

describe('<CreatePost> 페이지', () => {
  it('snapshot 체크', async () => {
    const component = render(<CreatePost />);
    await waitFor(() => expect(component.container).toMatchSnapshot());
  });
});
