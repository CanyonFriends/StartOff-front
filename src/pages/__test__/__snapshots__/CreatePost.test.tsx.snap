// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatePost> 페이지 snapshot 체크 1`] = `
<div>
  <div
    id="modal"
  />
  <header
    class="css-15jvef"
  >
    <div
      aria-label=""
      class="css-y8pvf0"
    >
      <div
        aria-label=""
        class="css-1jmbjs2"
      >
        <svg
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 0c-6.626 0-12 5.373-12 12 0 5.302 3.438 9.8 8.207 11.387.599.111.793-.261.793-.577v-2.234c-3.338.726-4.033-1.416-4.033-1.416-.546-1.387-1.333-1.756-1.333-1.756-1.089-.745.083-.729.083-.729 1.205.084 1.839 1.237 1.839 1.237 1.07 1.834 2.807 1.304 3.492.997.107-.775.418-1.305.762-1.604-2.665-.305-5.467-1.334-5.467-5.931 0-1.311.469-2.381 1.236-3.221-.124-.303-.535-1.524.117-3.176 0 0 1.008-.322 3.301 1.23.957-.266 1.983-.399 3.003-.404 1.02.005 2.047.138 3.006.404 2.291-1.552 3.297-1.23 3.297-1.23.653 1.653.242 2.874.118 3.176.77.84 1.235 1.911 1.235 3.221 0 4.609-2.807 5.624-5.479 5.921.43.372.823 1.102.823 2.222v3.293c0 .319.192.694.801.576 4.765-1.589 8.199-6.086 8.199-11.386 0-6.627-5.373-12-12-12z"
          />
        </svg>
      </div>
      <ul
        class="css-foxkse"
      />
    </div>
    <div
      class="css-12x73ce"
    >
      <div
        aria-label=""
        class="css-wysmox"
      >
        <a
          href="/signin"
        >
          <div
            aria-label=""
            class="css-ghj28t"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 9v-4l8 7-8 7v-4h-8v-6h8zm6-7c-1.787 0-3.46.474-4.911 1.295l.228.2 1.395 1.221c1.004-.456 2.115-.716 3.288-.716 4.411 0 8 3.589 8 8s-3.589 8-8 8c-1.173 0-2.284-.26-3.288-.715l-1.395 1.221-.228.2c1.451.82 3.124 1.294 4.911 1.294 5.522 0 10-4.477 10-10s-4.478-10-10-10z"
              />
            </svg>
          </div>
        </a>
      </div>
    </div>
  </header>
  <form
    class="css-30hrpu"
  >
    <div
      aria-label="post-title"
      class="css-1bechv"
    >
      <label
        aria-label=""
        class="css-1eeeix0"
        color="#000000"
        for="title"
      >
        제목
      </label>
      <input
        aria-label="title"
        class="css-1icg792"
        id="title"
        name=""
        placeholder="제목을 입력하세요"
        type="text"
        value=""
        width="100%"
      />
    </div>
    <div
      class="css-suin7r"
    >
      <h3
        class="css-1eseyvs"
        color="#000000"
      >
        현재 인원 / 모집 인원
      </h3>
      <div
        class="css-unaruz"
      >
        <div
          aria-label=""
          class="css-gd9wkw"
        >
          <button
            class="css-1fbvvo9"
            type="button"
            width="fit-content"
          >
            현재 인원 : 0
            <div
              aria-label=""
              class="css-4vb9q9"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 7.33l2.829-2.83 9.175 9.339 9.167-9.339 2.829 2.83-11.996 12.17z"
                />
              </svg>
            </div>
          </button>
        </div>
        <div
          aria-label=""
          class="css-gd9wkw"
        >
          <button
            class="css-1fbvvo9"
            type="button"
            width="fit-content"
          >
            모집 인원 : 0
            <div
              aria-label=""
              class="css-4vb9q9"
            >
              <svg
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 7.33l2.829-2.83 9.175 9.339 9.167-9.339 2.829 2.83-11.996 12.17z"
                />
              </svg>
            </div>
          </button>
        </div>
      </div>
    </div>
    <div
      class="css-0"
    >
      <div
        class="css-wbiz9q"
      >
        <div
          class="css-1knbux5"
        >
          <h3
            class="css-1eseyvs"
            color="#000000"
          >
            기술 스택
          </h3>
          <div
            aria-label=""
            class="css-gd9wkw"
          >
            <button
              class="css-1fbvvo9"
              type="button"
              width="fit-content"
            >
              스택 추가
              <div
                aria-label=""
                class="css-4vb9q9"
              >
                <svg
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 7.33l2.829-2.83 9.175 9.339 9.167-9.339 2.829 2.83-11.996 12.17z"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
        <ul
          class="css-kf5skc"
        />
      </div>
    </div>
    <div
      class="css-1diq458"
    >
      <header
        class="css-1tlhewg"
      >
        <span>
          Write
        </span>
        <span>
          Preview
        </span>
      </header>
      <div
        class="css-1ud59fp"
      >
        <div
          class="css-1hygdxi"
        >
          <textarea
            aria-label="markdown-textarea"
            class="css-1v6m50d"
            id=""
            name=""
            placeholder=""
          />
        </div>
        <div
          class="css-2n5s8x"
        >
          <span />
        </div>
      </div>
    </div>
    <button
      class="css-7rx0q1"
      type="submit"
      width="100%"
    >
      글 쓰기
    </button>
  </form>
</div>
`;
